File: benchmarks/coding-worker-v18/tasks/python-transactional-config/tests/verify_workspace.py
import hashlib
import json
import os
from pathlib import Path

root = Path('/workspace')
policy = json.loads(Path('/tests/workspace-policy.json').read_text())
current = {}
for directory, names, files in os.walk(root, followlinks=False):
    base = Path(directory)
    for name in tuple(names) + tuple(files):
        path = base / name
        if path.is_symlink():
            raise SystemExit('workspace link rejected')
    for name in files:
        path = base / name
        relative = path.relative_to(root).as_posix()
        content = path.read_bytes()
        if b'\x00' in content and relative != policy['binary_canary']:
            raise SystemExit('unexpected binary output')
        current[relative] = hashlib.sha256(content).hexdigest()
baseline = policy['baseline']
changed = set(baseline) ^ set(current)
changed.update(path for path in set(baseline) & set(current) if baseline[path] != current[path])
if len(changed) < policy['required_modified_files']:
    raise SystemExit('insufficient multi-file change')
canary = policy['binary_canary']
if current.get(canary) != baseline.get(canary):
    raise SystemExit('binary canary changed')

File: benchmarks/coding-worker-v18/tasks/typescript-command-palette-a11y/tests/verify_workspace.mjs
import { createHash } from 'node:crypto'
import { lstatSync, readFileSync, readdirSync } from 'node:fs'
import { join, relative, sep } from 'node:path'

const root = '/workspace'
const policy = JSON.parse(readFileSync('/tests/workspace-policy.json', 'utf8'))
const current = {}
function visit(directory) {
  for (const name of readdirSync(directory)) {
    const path = join(directory, name)
    const metadata = lstatSync(path)
    if (metadata.isSymbolicLink()) throw new Error('workspace link rejected')
    if (metadata.isDirectory()) { visit(path); continue }
    if (!metadata.isFile()) throw new Error('workspace special file rejected')
    const key = relative(root, path).split(sep).join('/')
    const content = readFileSync(path)
    if (content.includes(0) && key !== policy.binary_canary) throw new Error('unexpected binary output')
    current[key] = createHash('sha256').update(content).digest('hex')
  }
}
visit(root)
const names = new Set([...Object.keys(policy.baseline), ...Object.keys(current)])
const changed = [...names].filter(name => policy.baseline[name] !== current[name])
if (changed.length < policy.required_modified_files) throw new Error('insufficient multi-file change')
if (current[policy.binary_canary] !== policy.baseline[policy.binary_canary]) throw new Error('binary canary changed')
